### ISSUE TITLE
Second order derivative - Question about implementation 

### ISSUE THREAD

**Issue by JeremyAlain** [learnables/learn2learn#153]
Hi there, 
first of all I wanted to say that I really appreciate your framework. It's nicely written and really helpful. 

I have extensively gone through the whole code base to understand how you implemented MAML and have a qustion about the second_order implementation of MAML. Concretely I am refering to the omniglot example ``/master/examples/vision/maml_omniglot.py``. 

My problem is the understanding of how pytorch.clone() works in the context of your implementation. 
On line 94 you do 
```
learner = maml.clone()
batch = tasksets.train.sample()
evaluation_error, evaluation_accuracy = fast_adapt(batch,
                                                               learner,
                                                               loss,
                                                               adaptation_steps,
                                                               shots,
                                                               ways,
                                                               device)
```
You then go about calculating the second order derivatives for the learner object (if first_order=True) i.e. you do an update step for the current task (update \phi to \phi') and then evaluate on this task with the updated parameters \phi'. You then take the derivative of this loss  with respect to the original parameters.  You then repeat this process for all tasks. Notably for each task you do `learner=maml.clone()`. This all makes sense and is how the algorithm should work.

What I fail to understand is the following: 
On line 128 one can find the following code snippet: 
```
# Average the accumulated gradients and optimize
        for p in maml.parameters():
            p.grad.data.mul_(1.0 / meta_batch_size)
        opt.step()
```

But for each task we updated our learners and never actually touched the maml object i.e. updated the parameters of the maml object. So how does maml.parameters() even contain the accumulated gradients? I understand that this has probably to do with the way clone works. You even mention in the docstring of clone the following: 
"This implies that back-propagating losses on the cloned module will populate the buffers of the original module. For more information, refer to learn2learn.clone_module()." 

But somehow I don't understand how this exactly works. Because if updating a learner means that we directly also update the parameters of the maml object, this would imply that for each task when we do `learner=maml.clone()`we would actually clone parameters of the maml object that have been updated from the last task. Or put differently, it would mean that for each task we start from different initial parameters instead of starting off from the same parameters (as done in the paper). 

So somehow I believe that this cloning might have a more subtle meaning(maybe something with model._buffer). 

I hope that you might be able to explain what exactly is happening. 
Thanks

**Comment by seba-1511** [comment]
Hello @JeremyAlain, and thanks for the nice words.

Your understanding is mostly correct; the only missing piece is that the gradients in `maml.parameters()` are not populated with `clone()`, but by calling `evaluation_error.backward()`. This backward call backpropagates through the adaptation step, through the clone, and ends up in the parameters of the MAML model.

For a quick demonstration, see this colab link:
https://colab.research.google.com/drive/1OR9O3Zt6oyIV6Cn7QSDpNQevdk2XSdt8

Let me know if anything is unclear.
(This is a good question that came up a few times, so I'm going to use it as a reference.)

**Comment by JeremyAlain** [comment]
Hi @seba-1511, 
thanks so much for your quick answer. I played around with your colab and it helped a lot. I think it all makes sense now. I'll just try to recap it so that other people who might stumble upon this post have an elaborate answer. 

After cloning the maml model/object with `learner=maml.clone()` we do `fast_adapt`which in terms then calls `learner.adapt()`. In this function we call pytorch's `grad()` function. Note that this function calculates the gradients but does not directly apply them. We then use the function `maml_update`, which explicitely only updates the gradients of the clone learner with the just calculated gradient updates. We then use this updated learner to calculate the validation loss on the specific adaptation task. Back in the main function we then do `evaluation_error.backward()`which as explained above only calculates gradients for the original maml_model and not the clones. That means it also backpropagates through the clone (which is important to get the second order derivative, at least I think), but it then accumulates the gradient in the original maml object. It does this for all tasks, then takes the mean of the accumulated(i.e summed) gradients and then actually applies these gradients with `opt.step()`. It also means that the weights of the original maml object do not actually change until we do `opt.step(), thus for each task we start with the same weights when we clone a learner (which is how it should be)

In case you want to see my small extension of @seba-1511 colab you can view it here https://colab.research.google.com/drive/1_6CNv6_XmQmIIzyN33x7ewwRw5x6mRoS#scrollTo=QthBJF1zCwwD. I basically just went through the whole algorithm as defined in the omniglot tutorial and checked where and when the gradients of what get updated.
